flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java

```diff
@@ -0,0 +1,486 @@
+        if (!tempFolder.mkdir()) {
+            throw new IOException("Can not create temp folder: " + tempFolder);
+        }
+        outputStream = new BufferedOutputStream(new FileOutputStream(file));
+        indexStreams = new DataOutputStream[0];
+        dataStreams = new DataOutputStream[0];
+        indexFiles = new File[0];
+        dataFiles = new File[0];
+        lastValues = new byte[0][];
+        lastValuesLength = new int[0];
+        dataLengths = new long[0];
+        maxOffsetLengths = new int[0];
+        keyCounts = new int[0];
+    }
+
+    @Override
+    public void put(byte[] key, byte[] value) throws IOException {
+        int keyLength = key.length;
+
+        // Get the Output stream for that keyLength, each key length has its own file
+        DataOutputStream indexStream = getIndexStream(keyLength);
+
+        // Write key
+        indexStream.write(key);
+
+        // Check if the value is identical to the last inserted
+        byte[] lastValue = lastValues[keyLength];
+        boolean sameValue = lastValue != null && Arrays.equals(value, lastValue);
+
+        // Get data stream and length
+        long dataLength = dataLengths[keyLength];
+        if (sameValue) {
+            dataLength -= lastValuesLength[keyLength];
+        }
+
+        // Write offset and record max offset length
+        int offsetLength = VarLengthIntUtils.encodeLong(indexStream, dataLength);
+        maxOffsetLengths[keyLength] = Math.max(offsetLength, maxOffsetLengths[keyLength]);
+
+        // Write if data is not the same
+        if (!sameValue) {
+            // Get stream
+            DataOutputStream dataStream = getDataStream(keyLength);
+
+            // Write size and value
+            int valueSize = VarLengthIntUtils.encodeInt(dataStream, value.length);
+            dataStream.write(value);
+
+            // Update data length
+            dataLengths[keyLength] += valueSize + value.length;
+
+            // Update last value
+            lastValues[keyLength] = value;
+            lastValuesLength[keyLength] = valueSize + value.length;
+
+            valueCount++;
+        }
+
+        keyCount++;
+        keyCounts[keyLength]++;
+    }
+
+    @Override
+    public void close() throws IOException {
+        // Close the data and index streams
+        for (DataOutputStream dos : dataStreams) {
+            if (dos != null) {
+                dos.close();
+            }
+        }
+        for (DataOutputStream dos : indexStreams) {
+            if (dos != null) {
+                dos.close();
+            }
+        }
+
+        // Stats
+        LOG.info("Number of keys: {}", keyCount);
+        LOG.info("Number of values: {}", valueCount);
+
+        // Prepare files to merge
+        List<File> filesToMerge = new ArrayList<>();
+
+        try {
+
+            // Write metadata file
+            File metadataFile = new File(tempFolder, "metadata.dat");
```

Comment:
Why do the process including writing metadata file, building index file, statsing collisions, adding data files, and merging and writing to output execute in close? Is the close operation heavy?

Reply:
Closing just copy files into one file. It is not so heavy.